spccore/connection.py

```diff
@@ -0,0 +1,322 @@
+
+        :param download_requests: the list of download requests
+        :param use_multiple_threads: set to False to use single thread. Default True.
+        :return: a map between the DownloadRequest and the result
+        """
+
+
+def get_connection(*,
+                   repo_endpoint: str = SYNAPSE_DEFAULT_REPO_ENDPOINT,
+                   auth_endpoint: str = SYNAPSE_DEFAULT_AUTH_ENDPOINT,
+                   file_endpoint: str = SYNAPSE_DEFAULT_FILE_ENDPOINT,
+                   username: str = None,
+                   api_key: str = None
+                   ) -> SynapseConnection:
+    """
+    Get a connection to the Synapse server
+
+    :param repo_endpoint: the Synapse server repo endpoint
+    :param auth_endpoint: the Synapse server auth endpoint
+    :param file_endpoint: the Synapse server file endpoint
+    :param username: the Synapse username
+    :param api_key: the Synapse API key
+    :return: a Synapse connection
+    """
+    return SynapseConnection(repo_endpoint=repo_endpoint,
+                             auth_endpoint=auth_endpoint,
+                             file_endpoint=file_endpoint,
+                             username=username,
+                             api_key=api_key)
+
+
+# Helper functions
+
+def _generate_request_url(endpoint: str, request_path: str) -> str:
+    """
+    Generate the URL for the HTTP request
+
+    :param endpoint: the Synapse base endpoint
+    :param request_path: the unique path in the URI of the resource (i.e. "/entity/syn123")
+    :return: the URI of the resource
+    """
+    if endpoint is None or request_path is None:
+        raise ValueError("endpoint and request_path are required.")
+    if urllib_parse.urlparse(request_path).path != request_path:
+        raise ValueError('Incorrect format for request_path: {request_path}'.format(**{'request_path': request_path}))
+    return endpoint + request_path
+
+
+def _generate_signed_headers(url: str,
+                             *,
+                             username: str = None,
+                             api_key: str = None,
+                             headers: dict = None
+                             ) -> dict:
+    """
+    Generate headers signed with the API key
+
+    :param url: the URL used in the HTTP request
+    :param username: the user's username to sign the request
+    :param api_key: the user's API key to sign the request
+    :param headers: the HTTP request headers
+    :return: the signed headers
+    """
+
+    if url is None:
+        raise ValueError("url is required.")
+    if headers is None:
+        headers = dict(SYNAPSE_DEFAULT_HTTP_HEADERS)
+
+    headers = _enforce_user_agent(headers)
+
+    if username is None or api_key is None:
+        return headers
+
+    sig_timestamp = time.strftime(ISO_FORMAT, time.gmtime())
+    url = urllib_parse.urlparse(url).path
+    sig_data = username + url + sig_timestamp
+    signature = base64.b64encode(hmac.new(base64.b64decode(api_key),
```

Comment:
we can keep storing the api key as bytes to avoid having to do base64 decode for every request